aten/src/ATen/core/PythonFallbackKernel.cpp

```diff
@@ -82,14 +88,33 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
         if (nv.isNone()) {
           continue;
         }
-        auto* interpreter = nv.unsafeToTensorImpl()->pyobj_slot()->pyobj_interpreter();
-        if (interpreter) {
-          (*interpreter)->dispatch(op, stack);
-          return;
+
+        auto* t = nv.unsafeToTensorImpl();
+        if (t->key_set().has(c10::DispatchKey::Python)) {
+          tensors_with_python_key_present = true;
+        }
+
+        auto* t_interpreter = t->pyobj_slot()->pyobj_interpreter();
+        if (!interpreter && t_interpreter) {
+          interpreter = t_interpreter;
         }
       }
     }
   }
+
+  if (interpreter) {
+    if (tensors_with_python_key_present) {
+      (*interpreter)->dispatch(op, stack);
+    } else {
+      // At this point, there are no modes in the stack and no tensors with the python key.
+      // so remove and disable the python key before redispatching.
+      c10::impl::ExcludeDispatchKeyGuard no_python_guard(c10::DispatchKey::Python);
+      c10::DispatchKeySet keyset = dispatch_keys.remove(c10::DispatchKey::Python);
+      op.redispatchBoxed(keyset, stack);
```

Comment:
Is this equivalent to redispatch using `after_Python_keyset`?

Reply:
I may be wrong but I don't think so - IIUC `after_Python_keyset` is literally everything possible after Python, whereas we want to redispatch using only the *enabled* ones after Python.

Reply:
yep that's right (this says "take the exacted DispatchKeySet that the dispatcher computed before invoking the Python key, then remove the Python key and redispatch)